pkg/apis/config/v1alpha1/resourceinterpretercustomization_types.go

```diff
@@ -23,40 +25,226 @@ type ResourceInterpreterCustomization struct {
 	//             return desiredObj
 	//         end
 	//
-	// RetentionLua only holds the function body part, take the `desiredObj` and `observedObj` as the function
+	// LuaScript only holds the function body part, take the `desiredObj` and `observedObj` as the function
 	// parameters or global variables, and finally returns a retained specification.
 	// +required
-	RetentionLua string `json:"retentionLua"`
+	LuaScript string `json:"luaScript"`
+}
+
+// ReplicaResourceRequirement holds the scripts for getting the desired replicas
+// as well as the resource requirement of each replica.
+type ReplicaResourceRequirement struct {
+	// LuaScript holds the Lua script that is used to discover the resource's
+	// replica as well as resource requirements
+	// The script should implement a function as follows:
+	//     luaScript: >
+	//         function GetReplicas(desiredObj)
+	//             nodeClaim = {}
+	//             resourceRequest = {}
+	//             result = {}
+	//
+	//             result.replica = desiredObj.spec.replicas
+	//             result.resourceRequest = desiredObj.spec.template.spec.containers[0].resources.limits
+	//
+	//             nodeClaim.nodeSelector = desiredObj.spec.template.spec.nodeSelector
+	//             nodeClaim.tolerations = desiredObj.spec.template.spec.tolerations
+	//             result.nodeClaim = nodeClaim
+	//
+	//             return result
+	//         end
+	//
+	// LuaScript only holds the function body part, take the `desiredObj` as the function
+	// parameters or global variable, and finally returns the replica number and required resources.
+	// +required
+	LuaScript string `json:"luaScript"`
+}
+
+// ReplicaRevision holds the scripts for revising the desired replicas.
+type ReplicaRevision struct {
+	// LuaScript holds the Lua script that is used to revise replicas in the desired specification.
+	// The script should implement a function as follows:
+	//     luaScript: >
+	//         function ReviseReplica(desiredObj, desiredReplica)
+	//             desiredObj.spec.replicas = desiredReplica
+	//             return desiredObj
+	//         end
+	//
+	// LuaScript only holds the function body part, take the `desiredObj` and `desiredReplica` as the function
+	// parameters or global variables, and finally returns a revised specification.
+	// +required
+	LuaScript string `json:"luaScript"`
+}
+
+// StatusReflection holds the scripts for getting the status.
+type StatusReflection struct {
+	// LuaScript holds the Lua script that is used to get the status from the observed specification.
+	// The script should implement a function as follows:
+	//     luaScript: >
+	//         function ReflectStatus(observedObj)
+	//             status = {}
+	//             status.readyReplicas = observedObj.status.observedObj
+	//             return status
+	//         end
+	//
+	// LuaScript only holds the function body part, take the `observedObj` as the function
```

Comment:
Example A
``` lua
luaScript: >
 function ReflectStatus(observedObj)
     status = {}
     status.readyReplicas = observedObj.status.observedObj
     return status
 end
```
Example B
``` lua
luaScript: >
   status = {}
   status.readyReplicas = observedObj.status.observedObj
   return status

```

Which one is right?

Reply:
`Example` B is expected because it has been proved by @jameszhangyukun's demo.

But I'd prefer `Example A` which is easier to understand and straightforward.  @jameszhangyukun Can we load a whole function to Lua runtime instead of a piece of code?

Reply:
I think is Example B.

Reply:
> But I'd prefer `Example A` which is easier to understand and straightforward.

+1

Reply:
Maybe we can refer to nginx lua. 

https://github.com/openresty/lua-nginx-module#data-sharing-within-an-nginx-worker

Reply:
Can you make a demo with it?

Reply:
Great!, Let's do it by the function way, I'll update the API description later.